### PR TITLE
Remove APCu and replace with Redis instead

### PR DESCRIPTION
**Pull request by enoch85**
Fix https://github.com/nextcloud/vm/issues/2039

### CODE REVIEW

**Comment by szaimen** [comment]
Shouldn't we fix this in old installations via the update script, too? I have the feeling that this will fix some issues in the future, too...

**Comment by szaimen** [comment]
Question is then if switching from apcu to redis during the update will produce some issues, itself...

**Comment by enoch85** [comment]
> Question is then if switching from apcu to redis during the update will produce some issues itself...

We have to test this before of course.

I will fire up the latest release now just to see if it works. Can you do the same please? @szaimen

**Comment by szaimen** [comment]
What is still missing, afaics is switching from 'memcache.local' => '\\OC\\Memcache\\APCu', to 'memcache.local' => '\\OC\\Memcache\\Redis', in the config.php file during the update script...

File: nextcloud_update.sh [hunk 1]
```diff
@@ -324,13 +324,41 @@ then
     check_command phpenmod -v ALL redis
 fi
 
-# Upgrade APCu and igbinary
-#if [ "${CURRENTVERSION%%.*}" -ge "17" ]
-if [ "${CURRENTVERSION%%.*}" -ge "23" ]
+# Change from APCu to Redis for local cache
+if pecl list | grep redis >/dev/null 2>&1
+then
+    nextcloud_occ config:system:set memcache.local --value='\OC\Memcache\Redis'
```
Comment (enoch85) [inline nextcloud_update.sh:330]: @szaimen

Comment (szaimen) [reply nextcloud_update.sh:330]: ah sorry, didn't see this.
But shouldn't more likely be called
`nextcloud_occ config:system:set memcache.local --value='\\OC\\Memcache\\Redis'`

(notice the double slashes)

Comment (enoch85) [reply nextcloud_update.sh]: Nope :)

**Comment by enoch85** [comment]
Tested and works!

**Comment by enoch85** [comment]
Tested on multiple instances now, all is good. Nice!

**Comment by matthieu-labas** [comment]
I wanted to comment on issue #2039 but it's locked so I figured I can say it here: it *can* work with APCu if you add a `--define apc.enable_cli=1` to the php command line, whatever it is (in my case it was the cron).

From https://help.nextcloud.com/t/occ-wont-run-with-memcache-apcu/119724/3:

    sudo -u www-data php --define apc.enable_cli=1 -f "/opt/nextcloud/cron.php"

**Comment by enoch85** [comment]
@matthieu-labas I'm sure it works, but now we decided to not use it. We have been running Redis successfully for years, and then one of the core devs told us that APCu is better to run for local cache (yes it felt snappier) but then the major bug happened where it just stopped working (Nextcloud bug).

In other words, APCu as cache doesn't feel reliable, and we don't want to maintain something that is broken in Nextcloud, and constantly come up with workarounds to fix it. Redis on the other hand never failed, there are _0_ bug reports regarding it, and we made this VM since 2016.

**Comment by enoch85** [comment]
If you had issues with Redis, I really think you should try this: https://github.com/nextcloud/vm/blob/master/addons/redis-server-ubuntu.sh ;)

**Pull request merged**